sys/fmt/fmt.c

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ * @ingroup sys_fmt
+ * @{
+ * @file
+ * @brief String formatting library implementation
+ * @author Kaspar Schleiser <kaspar@schleiser.de>
+ * @}
+ */
+
+#include <assert.h>
+#include <stdarg.h>
+#include <stdint.h>
+#include <unistd.h>
+
+#include "div.h"
+#include "fmt.h"
+
+static const char _hex_chars[16] = "0123456789ABCDEF";
+
+static inline int _is_digit(char c)
+{
+    return (c >= '0' && c <= '9');
+}
+
+size_t fmt_byte_hex(char *out, uint8_t byte)
+{
+    if (out) {
+        *out++ = _hex_chars[byte >> 4];
+        *out = _hex_chars[byte & 0x0F];
+    }
+    return 2;
+}
+
+size_t fmt_strlen(const char *str)
+{
+    const char *tmp = str;
+    while(*tmp) {
+        tmp++;
+    }
+    return (tmp - str);
+}
+
+size_t fmt_str(char *out, const char *str)
+{
+    char c;
+    int len = 0;
+    if (!out) {
+        len = fmt_strlen(str);
+    } else {
+        while ((c = *str++)) {
+            *out++ = c;
+            len++;
+        }
+    }
+    return len;
+}
+
+size_t fmt_bytes_hex_reverse(char *out, const uint8_t *ptr, size_t n)
+{
+    size_t i = n;
+    while (i--) {
+        out += fmt_byte_hex(out, ptr[i]);
+    }
+    return (n<<1);
+}
+
+size_t fmt_u32_hex(char *out, uint32_t val)
+{
+    return fmt_bytes_hex_reverse(out, (uint8_t*) &val, 4);
```

Comment:
No idea how smart the compiler is here, but what about returning the literal `8` here instead of the result of this function call? The compiler may be able to cache this in a register for multiple calls. The same for `fmt_u64_hex` with the literal `16`.